### ISSUE TITLE
Read ROM command not working without serial delay

### ISSUE THREAD

**Issue by BizarroBull** [AndersBNielsen/Relatively-Universal-ROM-Programmer#7]
When a read command is sent to the arduino, it reads a block, begins serial connection with the host, writes the header byte 0xAA, writes the block, and ends serial communication.

However, for me the header byte is never sent correctly. I have to add a `delay(50);`.

Is that expected after the `readAddress(cAddr)` calls?
I see that delay call is already there for the burn rom command and talks about settling.

**Comment by AndersBNielsen** [comment]
I might be a bit sleepy but I can't seem to grasp what you are trying to do - 50ms is an eternity.  Are you using the read_binary.py script?

What exact steps are you taking to reproduce the issue?

**Comment by BizarroBull** [comment]
I've created a small java gui app and I'm doing the equivalent of read_binary.py. I send the bytes to my arduino to read the ROM.

The arduino receives my bytes properly, but when it writes 0xAA followed by the block of data, the 0xAA is received in my java app as garbage. I believe the block of data is received correctly but my app ignores it since it never received 0xAA. I think it's only the first byte that isn't sent correctly.

When I add the 50ms delay, the app receives the 0xAA and the block of data correctly.

Yes, 50 ms really slows it down. I haven't tried reducing the delay further. I was hoping you might have some insight to why it would be needed and what the minimum delay would need to be.

**Comment by AndersBNielsen** [comment]
Remember the Arduino resets when you open the port - that’s why it has a
long delay in the beginning.
Next you want to make sure you empty the serial buffer of all the junk
while waiting for 0xAA.
One byte at a time.

I assume you checked read_binary.py works with your Arduino and OS?
Serial kinda has a tendency to be handled differently :/

**Comment by BizarroBull** [comment]
I wait after opening the port so that's not the issue.
My java app is emptying the serial buffer but it never receives the 0xAA byte. It receives garbage in its place.
I haven't tried your read_binary.py script. I will install python and try that.

I was thinking this was an issue with sharing pins 15 and 16 of the Arduino. Aren't the other delays in your code related to that?

**Comment by BizarroBull** [comment]
I have the same problem using your python script instead of my java app. It works with my delay. It doesn't work without it.

Any thoughts?

**Comment by AndersBNielsen** [comment]
Where do you insert the delay?

**Comment by BizarroBull** [comment]
See the `delay(50);` statement in the code below:

```
      if (currentCommand.command == 0x01) {
        //Dump ROM via serial
        currentCommand.blockSize = Serial.read();
        byte cAddrL = Serial.read();
        byte cAddrH = Serial.read();
        cAddr = cAddrH << 8;
        cAddr |= cAddrL;
        romsize = (static_cast<uint32_t>(Serial.read()) << 8); //Reads stoppage a.k.a. the high byte of ROM size
        if (romsize == 0) romsize = 65536; //Need a fix to support A17+
        Serial.end();
        display.clearDisplay();
        display.print(F("Sending ROM via serial..."));
        display.print("Blocksize: ");
        display.print(currentCommand.blockSize);
        display.display();
        for (int i = 0; i < romsize/currentCommand.blockSize; i++) {
          // Read data into buffer
          for (uint16_t addr = 0; addr < currentCommand.blockSize; addr++) {
            buffer[addr] = readAddress(cAddr);
            cAddr++;
          }
          Serial.begin(baudrate);
          // delay(50); // Delay needed for Serial.write fix?
          Serial.write(0xAA);
          for (uint16_t addr = 0; addr < currentCommand.blockSize; addr++) {
            Serial.write(buffer[addr]);  
          }
          Serial.end();
        }
      }
```

**Comment by BizarroBull** [comment]
Would it have anything to do with the VEP (12.75V) and chip (ST 27C256) I'm using?

Is it worth watching the Arduino's D1 / TX pin with a logic analyzer while doing the readAddress followed by the Serial.begin/write?

**Comment by AndersBNielsen** [comment]
You actually DO need some sort of delay there - but on the order of 1ms,
not 50.
Your serial has to time out or read a byte and stay high long enough to be
registered as inactive for the start bit to be registered correctly.
Your chip should work, easily :)

Would be helpful if you double check what’s going on in hw

**Comment by BizarroBull** [comment]
I've moved the delay before the Serial.begin call. It works with a 1ms delay. Here you can see the read from the ROM chip, a 1 ms delay, then transmitting the data:

![rurp-1ms-delay](https://github.com/user-attachments/assets/97d53ea8-a264-410f-b462-8779da7d0905)

Without the delay the 0xAA looks like it's transmitted properly. My logic analyzer shows a framing error so the host receiving the data must have the same problem:

![rurp-no-delay](https://github.com/user-attachments/assets/0a96db91-4dae-40b9-8d0a-49e61733f28e)

I can't tell from the images where the Serial.begin call was. The voltage could still be changing at that point.

**Comment by AndersBNielsen** [comment]
Exactly. After transmitting junk it needs a whole frame of stop condition to know when to start. 
I guess this is cleared up and this can be closed - or is it missing somewhere it should be?

**Comment by BizarroBull** [comment]
The delay call needs to be added to your code.
In my comment above I show delay(50); as a comment after calling Serial.begin(baudrate).
I think it should be delay(1); BEFORE calling Serial.begin(baudrate).
